uom/common/container/kdtree.h

```diff
@@ -0,0 +1,627 @@
+  this->tree_deleted =
+      this->point_deleted && (!left || left->tree_deleted) && (!right || right->tree_deleted);
+}
+
+template <typename PointType>
+void KdTree<PointType>::TreeNode::PushDown() {
+  auto apply_to_child = [this](TreeNode* child) {
+    child->tree_deleted |= this->tree_deleted;
+    child->point_deleted |= child->tree_deleted;
+    if (this->tree_deleted) {
+      child->num_invalid_nodes = child->size;
+    }
+  };
+
+  if (left != nullptr) {
+    apply_to_child(left.get());
+  }
+  if (right != nullptr) {
+    apply_to_child(right.get());
+  }
+}
+
+template <typename PointType>
+double KdTree<PointType>::TreeNode::RangeTo(const PointType& query_point) const {
+  double squared_distance = 0;
+  for (int d = 0; d < kDim; ++d) {
+    const double val = query_point[d];
+    const double min_val = this->range.min()[d];
+    const double max_val = this->range.max()[d];
+    if (val < min_val) {
+      squared_distance += (val - min_val) * (val - min_val);
+    }
+    if (val > max_val) {
+      squared_distance += (val - max_val) * (val - max_val);
+    }
+  }
+  return std::sqrt(squared_distance);
+}
+
+template <typename PointType>
+void KdTree<PointType>::BuildTree(const std::vector<PointType>& points) {
+  if (points.empty()) {
+    return;
+  }
+
+  // Copy points as we need to sort the input points to build the tree.
+  std::vector<PointType> workspace = points;
+  auto workspace_span = absl::MakeSpan(workspace);
+  BuildTreeToNodeRecursive(workspace_span, &root_);
+}
+
+template <typename PointType>
+int KdTree<PointType>::FindBestSplitAxisAndSortByMaxSpan(absl::Span<PointType> points) {
+  CHECK_GT(points.size(), 0);
+
+  double min_value[kDim];  // NOLINT
+  double max_value[kDim];  // NOLINT
+  for (int d = 0; d < kDim; ++d) {
+    min_value[d] = points[0][d];
+    max_value[d] = points[0][d];
+  }
+  for (int i = 1; i < static_cast<int>(points.size()); ++i) {
+    for (int d = 0; d < kDim; ++d) {
+      min_value[d] = std::min(min_value[d], points[i][d]);
+      max_value[d] = std::max(max_value[d], points[i][d]);
+    }
+  }
+
+  int max_dim = 0;
+  double max_span = max_value[0] - min_value[0];
+  for (int d = 1; d < kDim; ++d) {
+    const double curr_span = max_value[d] - min_value[d];
+    if (curr_span > max_span) {
+      max_dim = d;
+      max_span = curr_span;
+    }
+  }
+
+  auto cmp = [max_dim](auto&& lhs, auto&& rhs) { return lhs[max_dim] < rhs[max_dim]; };
+  std::nth_element(points.begin(), points.begin() + points.size() / 2, points.end(), cmp);
+
+  return max_dim;
+}
+
+template <typename PointType>
+void KdTree<PointType>::BuildTreeToNodeRecursive(absl::Span<PointType> points,
+                                                 NodeUniquePtr* root_node) {
+  CHECK(root_node != nullptr);
+  if (points.empty()) {
+    root_node->reset();
+    return;
+  }
+
+  // Create new node if not exist.
+  if (*root_node == nullptr) {
+    *root_node = std::make_unique<TreeNode>();
```

Comment:
try [object pool](https://github.com/johnnysuns/common/blob/master/utils/memory/object_pool.h) :doge

Reply:
Good idea and LGTM for your object pool implementation. However, this PR is big enough,  let's use your object pool next PR. 😢

Reply:
BTW,  have you compare your object pool with boost::object_pool?